torch/csrc/profiler/collection.cpp

```diff
@@ -847,12 +853,76 @@ struct ResultGreater {
   }
 };
 
+void calculate_unique_tensor_ids(std::vector<result_ptr_t>& sorted_results) {
+  // Go through the result vectors
+  // For Torch Ops, check if (TensorImpl*, StorageImpl*) is already in the map
+  // For allocator events with negative allocation sizes,
+
+  // Do I need to track both TensorImpl and StorageImpl?
+
+  struct UniqueTensorInfo {
+    void* data_ptr_;
+    size_t unique_tensor_id_;
+  };
+
+  size_t cur_unique_tensor_id = 1;
+
+  std::unordered_set<void*> valid_storage_impls;
+  std::unordered_map<void*, UniqueTensorInfo> unique_tensor_map;
+
+  for (auto& res : sorted_results) {
+    if (auto torch_op =
+            c10::get_if<ExtraFields<EventType::TorchOp>>(&res->extra_fields_)) {
+      auto& metadata_vec = torch_op->inputs_.tensor_metadata_;
+      for (auto& metadata : metadata_vec) {
+        if (!metadata) {
+          continue;
+        }
+        auto tensor_impl = metadata->UNSAFE_tensor_impl_ptr_;
+
+        // This holds the storage pointer until we update it
+        void* data_ptr = (void*)metadata->unique_tensor_id_;
+        if (!tensor_impl || !data_ptr) {
+          continue;
+        }
+        if (valid_storage_impls.find(data_ptr) == valid_storage_impls.end()) {
+          cur_unique_tensor_id++;
+          valid_storage_impls.insert(data_ptr);
+          unique_tensor_map[tensor_impl] = {data_ptr, cur_unique_tensor_id};
+          metadata->unique_tensor_id_ = cur_unique_tensor_id;
+        }
+
+        auto it = unique_tensor_map.find(tensor_impl);
+        if (it == unique_tensor_map.end() || it->second.data_ptr_ != data_ptr) {
+          cur_unique_tensor_id++;
+          unique_tensor_map[tensor_impl] = {data_ptr, cur_unique_tensor_id};
+          metadata->unique_tensor_id_ = cur_unique_tensor_id;
+          continue;
+        }
+        metadata->unique_tensor_id_ = it->second.unique_tensor_id_;
+      }
+    } else if (
+        auto alloc_op = c10::get_if<ExtraFields<EventType::Allocation>>(
+            &res->extra_fields_)) {
+      if (alloc_op->alloc_size_ < 0) {
+        // For deallocations, clear relevant pointers
+        valid_storage_impls.erase(alloc_op->ptr_);
+        unique_tensor_map.erase(alloc_op->ptr_);
+      }
+    }
+  }
+}
+
 void build_tree(std::vector<std::shared_ptr<Result>>& events) {
   std::stable_sort(
       events.begin(), events.end(), [](const auto& a, const auto& b) {
         return a->start_time_ns_ < b->start_time_ns_;
       });
 
+  // Needs both TorchOp and Allocation events in order to apply
+  // the unique_tensor_id_ field, hence why it's done here.
+  calculate_unique_tensor_ids(events);
```

Comment:
nit: this feels like is should be called at the top level since it doesn't really have anything to do with tree building.